File: web/src/renderer/catalog.ts
import { createCatalog } from "@json-render/core";
import { z } from "zod";

const fileEntrySchema = z.object({
  name: z.string(),
  path: z.string(),
  is_dir: z.boolean(),
});

const sessionSchema = z.object({
  session_key: z.string(),
  agent: z.string().optional(),
  scope: z.string().optional(),
  purpose: z.string().optional(),
  summary: z.string().optional(),
  closed_at: z.string().optional(),
});

export const catalog = createCatalog({
  components: {
    Shell: {
      props: z.object({}),
      hasChildren: true,
    },
    Sidebar: {
      props: z.object({ slot: z.literal("sidebar").optional() }),
      hasChildren: true,
    },
    Main: {
      props: z.object({ slot: z.literal("main").optional() }),
      hasChildren: true,
    },
    Footer: {
      props: z.object({ slot: z.literal("footer").optional() }),
      hasChildren: true,
    },
    RightSidebar: {
      props: z.object({
        slot: z.literal("right").optional(),
        collapsed: z.boolean().optional(),
        onToggle: z.any().optional(),
        onOpenSettings: z.any().optional(),
      }),
      hasChildren: true,
    },
    FileTree: {
      props: z.object({
        entries: z.array(fileEntrySchema),
        childrenByPath: z.record(z.array(fileEntrySchema)),
        expanded: z.array(z.string()),
        selectedDir: z.string().nullable().optional(),
        rootId: z.string().nullable().optional(),
        managedRoots: z.array(z.string()).optional(),
      }),
    },
    DefaultListView: {
      props: z.object({ entries: z.array(fileEntrySchema) }),
    },
    FileViewer: {
      props: z.object({
        file: z
          .object({
            name: z.string(),
            path: z.string(),
            content: z.string(),
            encoding: z.string(),
            truncated: z.boolean(),
            size: z.number(),
            ext: z.string().optional(),
            mime: z.string().optional(),
          })
          .nullable()
          .optional(),
      }),
    },
    ActionBar: {
      props: z.object({
        status: z.string().optional(),
        pendingView: z.boolean().optional(),
        onAcceptView: z.any().optional(),
        onRevertView: z.any().optional(),
      }),
    },
    SessionList: {
      props: z.object({
        sessions: z.array(sessionSchema),
        selectedKey: z.string().optional(),
        onSelect: z.any().optional(),
      }),
    },
    SessionViewer: {
      props: z.object({
        session: sessionSchema.nullable().optional(),
      }),
    },
    SettingsPanel: {
      props: z.object({
        open: z.boolean().optional(),
      }),
    },
  },
  actions: {
    open: {
      params: z.object({ path: z.string(), root: z.string().optional() }),
    },
    open_dir: {
      params: z.object({ path: z.string(), root: z.string().optional() }),
    },
  },
});

File: server/internal/context/catalog.go
package context

func LoadCatalog() (ComponentCatalog, RegistrySchema) {
	catalog := ComponentCatalog{
		Version: "1.0",
		Components: map[string]ComponentSpec{
			"Shell":        {Description: "App shell layout", Props: map[string]any{"hasChildren": true}},
			"Sidebar":      {Description: "Left sidebar", Props: map[string]any{"slot": "sidebar"},},
			"Main":         {Description: "Main view area", Props: map[string]any{"slot": "main"}},
			"Footer":       {Description: "Footer area", Props: map[string]any{"slot": "footer"}},
			"RightSidebar": {Description: "Right sidebar", Props: map[string]any{"slot": "right", "collapsed": "boolean"}},
			"FileTree":     {Description: "File tree", Props: map[string]any{"entries": "FileEntry[]"}},
			"DefaultListView": {Description: "Default list", Props: map[string]any{"entries": "FileEntry[]"}},
			"FileViewer":   {Description: "File viewer", Props: map[string]any{"file": "File"}},
			"ActionBar":    {Description: "Action bar", Props: map[string]any{"status": "string"}},
			"SessionList":  {Description: "Session list", Props: map[string]any{"sessions": "Session[]"}},
			"SessionViewer": {Description: "Session viewer", Props: map[string]any{"session": "Session"}},
			"SettingsPanel": {Description: "Directory settings", Props: map[string]any{"open": "boolean"}},
		},
	}

	schema := RegistrySchema{
		"FileEntry": map[string]any{"name": "string", "path": "string", "is_dir": "boolean"},
		"File": map[string]any{
			"name": "string", "path": "string", "content": "string", "encoding": "string",
			"truncated": "boolean", "size": "number", "ext": "string", "mime": "string",
		},
		"Session": map[string]any{
			"session_key": "string", "agent": "string", "scope": "string", "purpose": "string",
			"summary": "string", "closed_at": "string",
		},
	}
	return catalog, schema
}
